### PR TITLE
Re-enable macOS CI to demonstrate PTY test hang

### PR DESCRIPTION
**Pull request by andrewleech**
## Re-enable macOS CI to demonstrate PTY test hang

The macOS runner was commented out because tests were freezing. This PR re-enables it to demonstrate the issue and start discussion on the best fix.

### What I Found

The hang happens in `test_pty_serial_read` when it calls `slave.flush()`. This triggers `tcdrain()` which has different behavior on macOS vs Linux when dealing with PTY devices:

- **Linux**: `tcdrain()` waits until data leaves the output buffer
- **macOS**: `tcdrain()` waits until someone *reads* the data from the other end

Since the test writes from the slave side before setting up a reader on the master, macOS blocks forever. This is a documented macOS kernel quirk (see pyserial/pyserial#625, python/cpython#97001).

### Potential Fixes

Two approaches to discuss:

#### Option 1: Fix the Test
Remove `slave.flush()` in `test_pty_serial_read` - it's not actually needed to verify data transmission.

**Pros:**
- Minimal change
- Test still validates PTY communication
- Zero risk to library behavior

**Cons:**
- Users calling `flush()` on macOS PTYs will still hit this
- Just papers over the library issue

#### Option 2: Fix the Library  
Make `flush()` detect PTY devices on macOS and skip the blocking `tcdrain()` call.

**Pros:**
- Fixes it for everyone using PTYs on macOS
- Matches Linux behavior
- Helps real use cases (terminal emulators, test frameworks, etc.)

**Cons:**
- More invasive change
- Need careful PTY detection logic
- Could affect semantics for edge cases

### This PR

This just re-enables the macOS runner so we can see the hang in CI and discuss which approach makes sense. The tests *will* fail/timeout - that's intentional to demonstrate the issue.

@zsquareplusc thoughts on which direction to take? I've got test branches for both approaches which fixes the problem but I'm not sure if there are "legitimate" usages of pty with the library outside of the unit test suite here that warrants fixing it in the library itself?

### CODE REVIEW

**Comment by projectgus** [comment]
FWIW Option 2 sounds reasonable to me. Thanks for digging into this, @andrewleech!

**Comment by kurtmckee** [comment]
@andrewleech I've implemented an "option 3" for expediency in #827. It allows the macOS test suite to pass by putting the call to `.read()` in a thread so that the call to `.tcdrain()` can succeed.

It may be ideal to pursue option #2 later macOS needs to be enabled in CI and #827 makes that happen without removing the call to `.flush()`.

Thanks for hunting down the root cause of this issue! :heart: :rocket:

**Comment by andrewleech** [comment]
I realized fixing the library is important, even just for the use case of pyserial based tools communicating with qemu which uses pty's.

I've got a working fix for this on my Mac, I just don't use it very often and it's got a flat battery so I haven't got back to it to push it to a PR!

**Comment by kurtmckee** [comment]
Excellent! I don't have a Mac so I need macOS running in CI to continue hammering on the test suite. Let's tackle this from both sides!

I'd like to see #827 and #817 get merged so I can continue the work of expanding the test suite to cover more code. :pray: :chart_with_upwards_trend:

**Comment by andrewleech** [comment]
Finally got my Mac charged up! Just pushed the library fix.

The approach: detect PTY devices on macOS (by checking if the path starts with `/dev/ttys*` or `/dev/pty[pqrs]*`) and skip the `tcdrain()` call entirely. Once `write()` completes, the data is in the kernel buffer, which is good enough for PTY semantics. Real serial hardware still gets the normal `tcdrain()` behavior.

This should fix the issue for anyone using pyserial with PTYs on macOS - not just our tests, but also folks communicating with QEMU or other emulators.

@kurtmckee your threading fix in #827 is a reasonable workaround for the test specifically, but this library fix means we don't need to restructure the test at all - the original test code with `flush()` now just works. Happy to discuss which approach (or both?) makes sense to merge.

File: serial/serialposix.py [hunk 1]
```diff
@@ -674,9 +674,38 @@ def flush(self):
...
+                          dev_name.startswith('/dev/ptyq') or
+                          dev_name.startswith('/dev/ptyr') or
+                          dev_name.startswith('/dev/ptys'))
+                if is_pty:
+                    # For PTY devices on macOS, tcdrain() blocks indefinitely
+                    # because it waits for data to be read from the other end,
+                    # which may never happen. Since the data is already in the
+                    # kernel buffer after write(), we consider flush() complete.
+                    return
+            except (OSError, IOError):
```
Comment (kurtmckee) [inline serial/serialposix.py]: `IOError` is an alias for `OSError`.

```suggestion
            except OSError:
```

**Comment by kurtmckee** [comment]
pyserial has code that checks for device symlinks when listing available ports. Is there ever a chance that sniffing for the device name could return a symlink instead of the underlying port?

Also, this currently has merge conflicts so CI isn't running against this branch yet.

**Comment by andrewleech** [comment]
> pyserial has code that checks for device symlinks when listing available ports. Is there ever a chance that sniffing for the device name could return a symlink instead of the underlying port?

I'm pretty sure `dev_name = os.ttyname(self.fd)` should be reliable here, it should return the real device path associated with the file descriptor, not the path that was used to open it.

I've just rebased, hopefully it works properly now!

**Comment by dpgeorge** [comment]
Reading through the linked issues for macOS and CPython, it seems that the root cause is a bug in CPython, where it does not release the GIL when doing a `tcdrain()`.  And it just so happens that Linux is OK with this but macOS is not.

It seems that CPython fixed this bug in https://github.com/python/cpython/pull/99503, which is back ported as far as CPython 3.10.

So... the work around added in this PR should not be needed for CPython 3.10 and above.  And the CI here uses CPython 3.10.11 and above.

Therefore, I'm confused what the real issue is.  Is there really still a bug in macOS kernel with `tcdrain()`?

**Comment by kurtmckee** [comment]
@dpgeorge The bug you linked to fixed a full-on interpreter lock caused by the failure to release the GIL. It didn't resolve the macOS kernel requirement that something read from the PTY before `tcdrain()` returns.

As demonstrated in #827, it's possible in CPython 3.10+ to have a secondary thread read from the PTY while the main thread calls `tcdrain()`, since 3.10+ now releases the GIL so that the reader thread is able to run at all.

The test suite fix in #827 would apparently have still hung on CPython 3.9, which I guess still locks the GIL and blocks all other threads.

**Comment by andrewleech** [comment]
> Therefore, I'm confused what the real issue is. Is there really still a bug in macOS kernel with `tcdrain()`?

There's definitely an issue still yes; the initial actions run for this PR failed in the same way I saw locally, with the test suite locking up during the test that used pty tcdrain.
https://github.com/pyserial/pyserial/actions/runs/19255227996/job/55048419958

Screenshot_20251205-072608

This first test run was with python 3.14.

@kurtmckee that sounds like a reasonable suggestion to the cause.  I've tried to use some AI research to deep dive this; came up with: https://github.com/apple/darwin-xnu/blob/2ff845c2e033bd0ff64b5b6aa6063a1f8f65aa32/bsd/kern/tty.c#L1657

  Confirmed: macOS PTY tcdrain() Behavior
  The XNU kernel's ttywait() function (called by TIOCDRAIN ioctl) waits until:
  1. tp->t_outq.c_cc (output queue character count) is zero, AND
  2. TS_BUSY flag is cleared

  For PTY devices specifically, the output queue doesn't drain until the master side reads the data. This differs from real serial hardware where data leaves the queue when transmitted to the wire.

  Sources Confirming This:
  1. https://github.com/pyserial/pyserial/issues/625 - Directly documents:
  "macOS interprets tcdrain() differently than Linux when dealing with PTY devices...macOS: tcdrain() waits until data is actually read from the other PTY endpoint"
  2. https://github.com/python/cpython/issues/97001 - The original CPython issue confirms the hang occurs specifically on macOS with PTYs, working fine on Linux.
  3. https://developer.apple.com/forums/thread/715119 - Documents the macOS-specific behavior (though the thread content wasn't fully extractable).
  4. https://github.com/apple/darwin-xnu/blob/main/bsd/kern/tty.c - Shows ttywait() loops while tp->t_outq.c_cc has data. For PTYs, this queue empties only when the master reads.

**Comment by dpgeorge** [comment]
OK, thanks for the extra info!
